lib/controllers/src/BasicControl.cpp

```diff
@@ -0,0 +1,78 @@
+#include "BasicControl.h"
+
+using namespace Controller;
+
+BasicControl::BasicControl(Actuator::Actuator *actuator, Encoder::Encoder *encoder,
+                           std::optional<Sensor::CurrentSensor *> currentSensor, PID::PID *pid, float maxDegPerSec,
+                           float maxCurrent, PinName lowerLimit, PinName upperLimit)
+    : m_actuator(actuator),
+      m_encoder(encoder),
+      m_currentSensor(currentSensor),
+      m_pid(pid),
+      m_maxDegPerSec(maxDegPerSec),
+      m_maxCurrent(maxCurrent),
+      m_lowerLimit(lowerLimit),
+      m_upperLimit(upperLimit) {}
+
+void BasicControl::stop() {
+  setSetPoint(0);
+  m_actuator->setValue(0);
+}
+
+void BasicControl::reset() {
+  stop();
+  m_encoder->reset();
+  if (m_currentSensor) {
+    m_currentSensor.value()->reset();
+  }
+  m_pid->reset();
+}
+
+std::optional<PID::PID *> BasicControl::getPID() {
+  return m_pid;
+}
+
+bool BasicControl::reportAngleDeg(float &angle) {
+  return m_encoder->getAngleDeg(angle);
+}
+
+bool BasicControl::reportAngularVelocityDegPerSec(float &speed) {
+  return m_encoder->getAngularVelocityDegPerSec(speed);
+}
+
+bool BasicControl::shouldUpdate() {
+  if (!m_ignoreRPMChecks.load()) {
+    float speed = 0;
+    if (!m_encoder->getAngularVelocityDegPerSec(speed)) {
+      return false;
```

Comment:
Kinda confused about the shouldUpdate() function, from my understanding, it's checking the RPM / current / limit switches if applicable, and if any of those are saturated then it returns that the controller should not update? Wouldn't that mean that if one of those attributes is saturated then the controller won't continue "updating" its actuator? How will the actuator ever get out of that saturated position?

Reply:
Also is getAngularVelocityDegPerSec() returning whether the encoder read was successful or not? That seems more like an issue with the measurement retrieval process instead of the retval of this shouldUpdate function

Reply:
Maybe I'm just not understanding the code here

Reply:
i think behaviour when limit check needs to depend on on what type of limit it is and what type of control it is. 

The old behaviour was if we were at position limit and the we were commanding it to go further (in vel or open loop), then we would send command of 0. if we were commanding it to reverse direction at the pos limit, we would respect that command. this type of logic may not hold in say position control mode for example

Reply:
For @cindyli-13: First Point: the idea is that the shouldUpdate() checks if the plant is staturated **_and_** it is being moved in the direction of saturation. Loosely, if we are saturated and getting worse, do nothing. It can get out of the saturated position of we ask it to go the other way. (At least this what I hope my code is saying).

Second Point: You are correct, getAngularVelocityDegPerSec(). And you bring up a good point. If a sensor/encoder reading fails it should just keep going. I'll address this is my next commit

Reply:
@wmmc88 Thats a good point. I hadn't thought of that. I'll try and figure out how to make it applicable for any control mode

Reply:
For encoder readings failing, a more robust solution is using the most up to date value if the reading fails (and setting some sort of fault that get communicated to sw). Also only doing that for a certain amount of time before its "dangerously old" and setting some other fault. Probably outside the scope of the pr but we should be implementing this sort of fault tracking and reporting eventually

Reply:
Yes rn we dont handle failures at all. we dont track, we dont count and we dont report. there will be a failure improvement task for each app at a later point. We can also discuss how we notify SW of failures during the CAN rearch meeting that we'll have post SAR

Reply:
@wmmc88 I think my latest push addresses ur concern. Basically, the logic is the same for all control modes: if the plant is saturated and getting worse or you're driving the plant too hard (rpm or current wise) then you stop. Then stop is defined differently for the different control modes. Basically Position control sets the setpoint as the current angle while the others set it to 0.

Reply:
If it's the position limit triggering and it's in position mode, the set point should probably be the limit and not the current value

Reply:
Not sure what the concern is. But to solve it the way you suggest, you'd have to pass in the angular position of the limit switches which makes the code more complicated. I think it's fair to assume that at any moment in time, the current angle will not be (considerably) beyond the limit switch.

Reply:
Position limit triggering as in based off abs enc reading. For limit switch triggering, current pos is fine

Reply:
we currently dont have a maxAngleTheta. I dont think we need something like. Limit switch trigerring is fine imo

Reply:
Fine for now maybe but limit switch failure is pretty common